ipa-core/src/protocol/ipa_prf/boolean_ops/share_conversion_aby.rs

```diff
@@ -90,56 +98,50 @@ pub(crate) enum Step {
 /// However, these terms are only non-zero when all `rs_{k}` terms are non-zero
 /// this happens with probability `1/(2^(256-m))` which is negligible for a sufficiently small `m`
 ///
+/// The implementation uses two type parameters to support vectorization. The type `XS` holds match
+/// keys. In the unvectorized case, `XS` is `AdditiveShare<BA64>`. In the vectorized case, `XS` is
+/// `BitDecomposed<AdditiveShare<BA{N}>>`. The type `YS` holds bitwise Fp25519 intermediates. In the
+/// unvectorized case, `YS` is `AdditiveShare<BA256>`. In the vectorized case, `YS` is
+/// `BitDecomposed<AdditiveShare<BA{n}>>`.
+///
 /// # Errors
 /// Propagates Errors from Integer Subtraction and Partial Reveal
-pub async fn convert_to_fp25519<C, B>(
+pub async fn convert_to_fp25519<C, XS, YS, const N: usize>(
     ctx: C,
     record_id: RecordId,
-    x: &AdditiveShare<B>,
-) -> Result<AdditiveShare<Fp25519>, Error>
+    x: XS,
+) -> Result<AdditiveShare<Fp25519, N>, Error>
 where
     C: Context,
-    B: SharedValue + CustomArray<Element = Boolean>,
+    Fp25519: Vectorizable<N>,
+    Boolean: FieldSimd<N>,
+    XS: ArrayAccessRef<Element = AdditiveShare<Boolean, N>>,
+    YS: ArrayAccessRef<Element = AdditiveShare<Boolean, N>>
+        + ArrayBuild<Input = AdditiveShare<Boolean, N>>
+        + FromPrss<usize>,
+    AdditiveShare<Boolean, N>: SecureMul<C>
+        + Reveal<C, N, Output = <Boolean as FieldVectorizable<N>>::ArrayAlias>
+        + Not<Output = AdditiveShare<Boolean, N>>,
+    Vec<AdditiveShare<BA256>>: for<'a> TransposeFrom<&'a YS>,
+    Vec<BA256>:
+        for<'a> TransposeFrom<&'a [<Boolean as Vectorizable<N>>::Array; 256], Error = Infallible>,
```

Comment:
I agree that trait bounds are becoming concerning - is there a way to get them under an umbrella trait?

Reply:
One pretty easy change is to add a `BooleanProtocols` umbrella trait (similar to `BasicProtocols`). I just made the code changes, but I'll push it in a separate PR just to avoid any more late-arriving changes on this one.

Besides that, the main way I can see to simplify the trait bounds is to remove the ability to be generic over different vectorization dimensions -- or at least remove the ability to use a compact representation (`BA{n}` instead of `[Boolean; {n}]`) when the vectorization dimension is 1.

Reply:
I like this direction - it is not worth maintaining compatibility with `BA{n}` if it costs that much imo